### PR TITLE
Log a more helpful message when no candidates are given

### LINKED ISSUE #1469: Test logs full of: 'unable to allocate a target store for {[] }'
A test run containing at least one failure is quite full of:
```
E0623 17:45:25.730869   52414 queue.go:266] failure processing range range=1 ["" - "\xff\xff") from replicate queue: allocator.go:207: unable to allocate a target store for {[] }
```
This is weird, because it seems to imply that `storage.allocator.allocateTargetInternal` is being called with an empty array passed for `existing`. Looks like a bug to me. @spencerkimball?

### PR DESCRIPTION
**Pull request by tamird**
Closes #1469.

### CODE REVIEW

**Comment by tamird** [comment]
@bdarnell PTAL

File: storage/allocator.go [hunk 1]
```diff
@@ -169,8 +169,10 @@ func (a *allocator) AllocateTarget(required proto.Attributes, existing []proto.R
 
 func (a *allocator) allocateTargetInternal(required proto.Attributes, existing []proto.Replica,
 	relaxConstraints bool, filter func(*proto.StoreDescriptor, *stat, *stat) bool) (*proto.StoreDescriptor, error) {
-	attrs := append([]string(nil), required.Attrs...)
-	for {
+	// Because more redundancy is better than less, if relaxConstraints, the
+	// matching here is lenient, and tries to find a target by relaxing an
+	// attribute constraint, from last attribute to first.
+	for attrs := append([]string(nil), required.Attrs...); ; attrs = attrs[:len(attrs)-1] {
```
Comment (bdarnell) [inline storage/allocator.go]: I find this less clear than the old version; I think the three-argument `for` statement only adds clarity when the second argument is used.

File: storage/allocator.go [hunk 2]
```diff
@@ -198,15 +200,11 @@ func (a *allocator) allocateTargetInternal(required proto.Attributes, existing [
...
-
-		// Otherwise, we have not found a store. Because more redundancy
-		// is better than less, if relaxConstraints, we still try to find
-		// a target by relaxing an attribute constraint, from last
-		// attribute to first.
-		if len(attrs) == 0 || !relaxConstraints {
+		if len(attrs) == 0 {
+			return nil, util.Errorf("unable to allocate a target store; no candidates available")
+		} else if !relaxConstraints {
 			return nil, util.Errorf("unable to allocate a target store for %s", required)
```
Comment (bdarnell) [inline storage/allocator.go]: I think the message "no candidates available" helps whether there are attributes or not; how about making the second form "unable to allocate a target store; no candidates available with attributes %s".

**Comment by bdarnell** [comment]
LGTM

File: storage/allocator.go (hunk 2 above)
Comment (tschottdorf) [inline storage/allocator.go]: Could you give `proto.Attributes` a decent version of `String()`?

**Comment by tschottdorf** [comment]
LGTM

File: storage/allocator.go (hunk 1 above)
Comment (tschottdorf) [inline storage/allocator.go]: seconded.

File: storage/allocator.go (hunk 2 above)
Comment (tamird) [inline storage/allocator.go]: Sure, how about I delegate to `SortedString`?

Comment (tschottdorf) [inline storage/allocator.go]: order matters, unfortunately.

Comment (tschottdorf) [inline storage/allocator.go]: wait, not true. Sure, `SortedString` looks good.

Comment (tamird) [inline storage/allocator.go]: order does kinda matter because of how attributes get stripped off one at a time. I'll do the ordered thing

Comment (tamird) [inline storage/allocator.go]: OK PTAL

**Pull request merged**